components/core/src/reducer/ServerContext.hpp

```diff
@@ -0,0 +1,184 @@
+#ifndef REDUCER_SERVERCONTEXT_HPP
+#define REDUCER_SERVERCONTEXT_HPP
+
+#include <optional>
+#include <set>
+
+#include <boost/asio.hpp>
+#include <json/single_include/nlohmann/json.hpp>
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+
+#include "../clp/TraceableException.hpp"
+#include "CommandLineArguments.hpp"
+#include "Pipeline.hpp"
+
+namespace reducer {
+enum class ServerStatus : uint8_t {
+    Idle,
+    Running,
+    ReceivedAllResults,
+    RecoverableFailure,
+    UnrecoverableFailure
+};
+
+namespace cJobAttributes {
+constexpr char JobId[] = "job_id";
+constexpr char BucketSize[] = "bucket_size";
+}  // namespace cJobAttributes
+
+/**
+ * Class which manages interactions with the jobs database and result cache database. Also holds
+ * state for the reducer job this server is handling.
+ */
+class ServerContext {
+public:
+    // Types
+    class OperationFailed : public clp::TraceableException {
+    public:
+        // Constructors
+        OperationFailed(clp::ErrorCode error_code, char const* filename, int line_number)
+                : clp::TraceableException(error_code, filename, line_number) {}
+
+        // Methods
+        [[nodiscard]] char const* what() const noexcept override {
+            return "ServerContext operation failed";
+        }
+    };
+
+    // Constructors
+    explicit ServerContext(CommandLineArguments& args);
+
+    // Methods
+    /**
+     * Resets all state in the ServerContext. This method must be called between invocations of
+     * run().
+     */
+    void reset();
```

Comment:
Instead of a dedicated `reset` method, could we just recreate the ServerContext object in the main loop?

Reply:
The reason I do it this way instead of recreating the context is because it eliminates timing issues with the port being available for the reducer's socket -- had some issues when doing it the way you suggest where the OS wasn't ready to give the port back by the time the reducer was getting ready for the next job